### DISCUSSION TITLE
Can we reuse the glb model loaded by gltfio

### DISCUSSION THREAD

**Discussion by yicong123456** [google/filament#7627] [Q&A]
I used gltfio to load the glb model and add them to the scene and then transform it to the position using transformManager. I need to create multiple objects using the same glb model and transform them to different position in the scene, and it seems that I need load glb modle n times if I want to add n object using the same model which will have bad performance. May I know if there are some proper ways to achieve this such that I can reuse the assets got from model like reuse the same material by materialInstance while rendering. Thanks so much!

**Comment by nmfisher** [comment]
You can create instances of the same underlying asset that share vertex buffers/textures/etc but have their own transforms. Look at `AssetLoader::createInstancedAsset`.

**Reply by yicong123456** [reply]
Oh, that's awesome! I read the document and suppose I should could achieve using createInstancedAsset, and I tried like these:


createInstancedAsset: 
               val position2 = Float3.obtain().setData(world.vehicle.position.x.toFloat(), world.vehicle.position.y.toFloat(), 2.3f)
               if(!this::vehicleAsset.isInitialized) {
                   vehicleAsset = assetLoader.createInstancedAsset(buffer, arrayOf(asset.instance))!!
               }
               if (vehicleAsset != null) {
                   addModelToScene(vehicleAsset, position2)
               }

origin:
             val position1 = Float3.obtain().setData(world.vehicle.position.x.toFloat(), world.vehicle.position.y.toFloat(), 2.3f)
             addModelToScene(asset, position1)

the addModelToScene method do nothing except receiving the FilamentAsset and add it to the scene and then set transform to it:
      
       private fun addModelToScene(asset: FilamentAsset, pos: Float3) {
        if(!assetSet.contains(asset.root)) {
            scene.addEntities(asset?.entities)
            assetSet.add(asset.root)
        }

        val transformManager = engine.transformManager
        val rootTransform = transformManager.getInstance(asset.root)

        val matrix = FloatArray(16)
        transformManager.getTransform(rootTransform, matrix)
        val modelMatrix = FloatArray(16)
        Matrix.setIdentityM(modelMatrix, 0)

        val translateMatrix = FloatArray(16)
        Matrix.setIdentityM(translateMatrix, 0)
        Matrix.translateM(translateMatrix, 0, pos.x, pos.y, pos.z)
        Matrix.multiplyMM(modelMatrix, 0, modelMatrix, 0, translateMatrix, 0)

        val scaleMatrix = FloatArray(16)                                                                                                                      
        Matrix.setIdentityM(scaleMatrix, 0)
        Matrix.scaleM(scaleMatrix, 0, this.tmpV.toFloat(), (this.tmpV+0.0067).toFloat(), (this.tmpV).toFloat())
        Matrix.multiplyMM(modelMatrix, 0, modelMatrix, 0, scaleMatrix, 0)

        val rotateMatrix = FloatArray(16)
        Matrix.setIdentityM(rotateMatrix, 0)
        Matrix.rotateM(rotateMatrix, 0, theta.toFloat(), 0f, 0f, 1f)
        Matrix.multiplyMM(modelMatrix, 0, modelMatrix, 0, rotateMatrix, 0)

        val rotateMatrix1 = FloatArray(16)
        Matrix.setIdentityM(rotateMatrix1, 0)
        Matrix.rotateM(rotateMatrix1, 0, theta1.toFloat(), 0f, 1f, 0f)
        Matrix.multiplyMM(modelMatrix, 0, modelMatrix, 0, rotateMatrix1, 0)

        val rotateMatrix2 = FloatArray(16)
        Matrix.setIdentityM(rotateMatrix2, 0)
        Matrix.rotateM(rotateMatrix2, 0, theta2.toFloat(), 1f, 0f, 0f)
        Matrix.multiplyMM(modelMatrix, 0, modelMatrix, 0, rotateMatrix2, 0)

        transformManager.setTransform(rootTransform, modelMatrix)
    }

I can confirm the addModelToScene method executed with the same position value for both cases. But I can get the render result for the original data but get nothing for the asset I created using createInstancedAsset. Did I miss anything or make any mistake while achieving this and thanks so much for your kindly help!

**Comment by nmfisher** [comment]
This looks like the Kotlin API, which I'm not 100% familiar with, but you probably need to add the instance entities to the scene, not the asset entities (pseudocode):
```
for(instance in instances)   
         scene.addEntities(instance.entities)
}
```

**Reply by yicong123456** [reply]
Yes, I tried to use assetLoader.createInstance and used the FilamentInstance instead of FilamentAssset and it finally worked. Thanks so much for your help!!!
